### PR TITLE
Feature/machine state

### PR DESCRIPTION
**Pull request by LarissaKoehne**

### CODE REVIEW

File: src/web/Scenarios/MachineState/MachineStateService.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            if (string.IsNullOrWhiteSpace(machineId))
+            {
+                throw new ArgumentException("Value cannot be null or whitespace.", nameof(machineId));
+            }
+
+            using (_logger.BeginScope(new Dictionary<string, object>{ { "MachineId", machineId }, { "RequestEndpoint", TapioMachineStateEndpoint } }))
+            {
+                _logger.LogInformation("Receiving states");
+                var token = await _tokenProvider.ReceiveTokenAsync(TapioScope.CoreApi);
+                HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, MachineStateRequest);
```
Comment (SebastianRoether) [inline src/web/Scenarios/MachineState/MachineStateService.cs]: var

File: src/web/Scenarios/MachineState/MachineStateService.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        private readonly ILogger<MachineStateService> _logger;
+
+        public MachineStateService(HttpClient httpClient, ITokenProvider tokenProvider, ILogger<MachineStateService> logger)
+        {
+            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            _tokenProvider = tokenProvider ?? throw new ArgumentNullException(nameof(tokenProvider));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<JToken> ReceiveStateOfSingleMachineAsync(string machineId, CancellationToken cancellationToken)
```
Comment (SebastianRoether) [inline src/web/Scenarios/MachineState/MachineStateService.cs]: The service shouldn't return a JToken. Please fix this after fixing everything else, I may even accept this PR and have this as bug with a separate pr

File: src/web/Startup.cs [hunk 3]
```diff
@@ -1,104 +1,117 @@
...
+                {
+                    builder.AddSerilog(dispose: true);
+                })
+                .AddSingleton<IScenarioCrawler, ScenarioCrawler>()
+                .AddSingleton<IScenarioRepository, ScenarioRepository>()
+                .AddSingleton<ITokenProvider, TokenProvider>()
+                .AddSingleton<OptionsValidator>()
+                .AddMvc()
+                .SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
+            services
```
Comment (SebastianRoether) [inline src/web/Startup.cs]: isnt AddHttpClient also fluent? why dont we write
```
   .SetCompatibilityVersion
   .AddHttpClient()
   .AddHttpClient()
   ....
   .AddOptions<>
```

File: src/web/src/app/app.component.html [hunk 4]
```diff
@@ -2,9 +2,9 @@
   class="main-header level">
   <div class="level-left">
     <div class="level-item">
-      <a href="https://tapio.one/" target="_blank" rel="noopener"><img src="assets/img/logo/tapio.svg" class="logo"></a>
-      <span> developer app by </span>
-      <a href="https://aitgmbh.de/" target="_blank" rel="noopener"><img src="assets/img/logo/ait.svg" class="logo"></a>
+      <a href="https://tapio.one/" target="_blank" rel="noopener" class="is-hidden-mobile"><img src="assets/img/logo/tapio.svg" class="logo" alt="tapio"></a>
```
Comment (SebastianRoether) [inline src/web/src/app/app.component.html]: This again changes the app component header. Is that intended?

File: src/web/src/app/scenario-historicconditions/scenario-historicconditions.component.html [hunk 5]
```diff
@@ -107,15 +106,14 @@
...
-    <div>
-      No data available for this machine
-    </div>
+  <div *ngIf="(error$ | async); else loading" class="errorMessage">
+    No data available for this machine
   </div>
-  <div>
-    <div *ngIf="(loading$ | async)" class="spinner">
+
+  <ng-template #loading>
```
Comment (SebastianRoether) [inline src/web/src/app/scenario-historicconditions/scenario-historicconditions.component.html]: no async after loading$ looks suspicious. Can you make sure this template is the same as in the other story and works the same?

File: src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.html [hunk 6]
```diff
@@ -23,16 +23,14 @@ <h1 class="card-header-title">{{subscription.name}}</h1>
...
-    <div *ngIf="errorLoading$ | async; else loading" class="message is-danger">
-      <div class="message-body">
-        An Error encountered in loading the subscriptions. Please try again later.
-        <p class="is-size-7">Consult the developer console for more information</p>
-      </div>
+  <div *ngIf="(error$ | async); else loading" class="errorMessage">
+    No data available for this machine
+  </div>
+
+  <ng-template #loading>
```
Comment (SebastianRoether) [inline src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.html]: same for this template

File: src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.ts [hunk 7]
```diff
@@ -5,21 +5,20 @@ import { Subscription } from "../shared/models/subscription.model";
...
 @Component({
   selector: "app-scenario-machineoverview",
-  templateUrl: "./scenario-machineoverview.component.html",
-  styleUrls: ["./scenario-machineoverview.component.css"]
+  templateUrl: "./scenario-machineoverview.component.html"
 })
 export class ScenarioMachineoverviewComponent implements OnInit {
     subscriptions$: Observable<Subscription[]>;
-    errorLoading$ = new Subject<boolean>();
+    error$ = new Subject<boolean>();
```
Comment (SebastianRoether) [inline src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.ts]: Ahh...thats probably what confuses me. Suffixing a variable with ``$`` is only best practice for ``observable``. But this will also be a different bug with different PR

File: src/web/src/app/scenario-machinestate/detail/detail.component.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    ngOnInit(): void {
+        const defaultMessage = "No data to display";
+
+        this.id$.subscribe(id => {
+            if (!id){
+                return;
+            }
+
+            this.machineStateService.getLastKnownStateFromMachine(id).subscribe(
```
Comment (SebastianRoether) [inline src/web/src/app/scenario-machinestate/detail/detail.component.ts]: Need to save subscription in private variable and unsubscribe here if that variable is set - before creating a new subscription

File: src/web/src/app/scenario-machinestate/detail/detail.component.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        private readonly machineStateService: MachineStateService,
+        private readonly route: ActivatedRoute,
+        public readonly domSanitizerService: DomSanitizer
+    ) {}
+
+    ngOnInit(): void {
+        const defaultMessage = "No data to display";
+
+        this.id$.subscribe(id => {
+            if (!id){
```
Comment (SebastianRoether) [inline src/web/src/app/scenario-machinestate/detail/detail.component.ts]: Not so sure about that, but cant you use a Subject instead of a BehaviorSubject? Then you wouldnt need this null-check i guess

File: src/web/src/app/scenario-sample/scenario-sample.component.ts [hunk 10]
```diff
@@ -10,13 +11,13 @@ import { HelloWorldService } from "./scenario-sample.service";
...
 })
 export class ScenarioSampleComponent implements OnInit {
-    public Welcome: string;
+    public Welcome$: Observable<string>;
 
-    constructor(private helloWorldService: HelloWorldService) { }
+    constructor(private readonly helloWorldService: HelloWorldService) { }
 
     ngOnInit(): void {
         this.helloWorldService.getWelcome().subscribe(welcome => {
```
Comment (SebastianRoether) [inline src/web/src/app/scenario-sample/scenario-sample.component.ts]: Isn't ``this.Welcome$ = this.helloWorldService.getWelcome()`` easier? Then we wouldn't have to deal with unsubscribing which is missing here

File: src/web.Tests.Unit/HttpMessageHandlerVerificationExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+using Moq;
+using Moq.Protected;
+
+namespace Aitgmbh.Tapio.Developerapp.Web.Tests.Unit
+{
+    public static class HttpMessageHandlerVerificationExtensions
+    {
+        private const string SendAsyncMethodName = "SendAsync";
+
+        public static void VerifySendAsyncWasInvokedExactlyOnce(this Mock<HttpMessageHandler> instance)
```
[the remark this answers is not in the document]
Comment (LarissaKoehne) [reply src/web.Tests.Unit/HttpMessageHandlerVerificationExtensions.cs]: Yes, I obviously didn't see this one. I'll delete the new one I just created.

File: src/web/Startup.cs (hunk 3 above)
Comment (LarissaKoehne) [reply src/web/Startup.cs]: I tried it, but that doesn't seem to work with AddHttpClient...

File: src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.ts (hunk 7 above)
Comment (LarissaKoehne) [reply src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.ts]: But a Subject is an observable: http://reactivex.io/rxjs/manual/overview.html#subject

File: src/web/src/app/scenario-sample/scenario-sample.component.ts (hunk 10 above)
Comment (LarissaKoehne) [reply src/web/src/app/scenario-sample/scenario-sample.component.ts]: Thank you - that was code I haven't touched though.

File: src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.ts (hunk 7 above)
Comment (SebastianRoether) [reply src/web/src/app/scenario-machineoverview/scenario-machineoverview.component.ts]: Okay, there seems to be no real naming convention. So having a dollar sign at both is okay

File: src/web/Startup.cs (hunk 3 above)
Comment (SebastianRoether) [reply src/web/Startup.cs]: ok
